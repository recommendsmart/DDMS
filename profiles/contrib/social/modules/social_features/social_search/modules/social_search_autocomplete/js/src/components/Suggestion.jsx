import React from "react";
import PropTypes from "prop-types";

/**
 * Escapes a string for use as literal RegExp value.
 *
 * @param {string} string
 *   The user string to escape.
 * @return {string}
 *   The string with all special characters escaped.
 */
function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"); // $& means the whole matched string
}

/**
 * Returns the target string with <b> tags around the highlight word.
 *
 * @param {string} highlight
 *   The word(s) to highlight in the target.
 * @param {string} target
 *   The string to process for highlighting.
 * @return {array}
 *   An array to make up the string that will properly be interpreted by React.
 */
function highlightText(highlight, target) {
  // Convert the highlight string to a RegExp value that is properly escaped and
  // can be used to match all words.
  const highlightWords = escapeRegExp(highlight.trim()).replace(/\s/, '|');

  // Create a capturing regular expression for splitting. This ensures that when
  // we iterate we don't lose the search string.
  const search = new RegExp(`(${highlightWords})`, "gi");
  const parts = target.split(search);

  // Assemble the string again, highlighting our search term.
  const text = [];

  // eslint-disable-next-line
  for (const i in parts) {
    if (search.test(parts[i])) {
      text.push(<b key={i}>{parts[i]}</b>);
    } else {
      text.push(parts[i]);
    }
  }

  return text;
}

function Suggestion(props) {
  const { query, type, label, url, tags } = props;

  const highlightedLabel = highlightText(query, label);

  const markupTags = [];

  // eslint-disable-next-line
  for (const i in tags) {
    markupTags.push(
      <span
        key={i}
        className="search-suggestion__tag badge badge-default badge--pill"
      >
        {tags[i]}
      </span>
    );
  }

  return (
    <a className="search-suggestion" href={url} tabIndex="0">
      <div className="search-suggestion__type">{type}</div>
      <div className="search-suggestion__body">
        <div className="search-suggestion__label">{highlightedLabel}</div>
        {tags.length ? (
          <div className="search-suggestion__tags">{markupTags}</div>
        ) : null}
      </div>
    </a>
  );
}

Suggestion.propTypes = {
  query: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
  tags: PropTypes.arrayOf(PropTypes.string),
};

Suggestion.defaultProps = {
  tags: [],
};

export default Suggestion;
