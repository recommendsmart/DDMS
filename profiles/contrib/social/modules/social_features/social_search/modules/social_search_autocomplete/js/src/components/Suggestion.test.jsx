import React from "react";
import renderer from "react-test-renderer";
import Suggestion from "./Suggestion";

describe("Suggestion", () => {
  it("renders a topic suggestion", () => {
    const tree = renderer
      .create(
        <Suggestion
          query="open"
          type="Topic"
          label="Open Social is great for customisable communities"
          url="http://example.com/node/1"
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders an event with a tag", () => {
    const tree = renderer
      .create(
        <Suggestion
          query="launch"
          type="Event"
          label="Open Social launch party"
          url="http://example.com/node/2"
          tags={["You are enrolled to this event"]}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
