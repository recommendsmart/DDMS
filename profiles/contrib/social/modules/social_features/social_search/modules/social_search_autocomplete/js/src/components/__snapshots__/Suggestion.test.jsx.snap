// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Suggestion renders a topic suggestion 1`] = `
<a
  className="search-suggestion"
  href="http://example.com/node/1"
  tabIndex="0"
>
  <div
    className="search-suggestion__type"
  >
    Topic
  </div>
  <div
    className="search-suggestion__body"
  >
    <div
      className="search-suggestion__label"
    >
      
      <b>
        Open
      </b>
       Social is great for customisable communities
    </div>
  </div>
</a>
`;

exports[`Suggestion renders an event with a tag 1`] = `
<a
  className="search-suggestion"
  href="http://example.com/node/2"
  tabIndex="0"
>
  <div
    className="search-suggestion__type"
  >
    Event
  </div>
  <div
    className="search-suggestion__body"
  >
    <div
      className="search-suggestion__label"
    >
      Open Social 
      <b>
        launch
      </b>
       party
    </div>
    <div
      className="search-suggestion__tags"
    >
      <span
        className="search-suggestion__tag badge badge-default badge--pill"
      >
        You are enrolled to this event
      </span>
    </div>
  </div>
</a>
`;
